Lab Assignment 1

~ Rhythm Shah

~ 1233960561

~ 01/25/2025

## Code Cell 1 (5%) :  Library and data import. Show the summary of the input data.

In [1]:
import pandas as pd
import spacy
import nltk
from spacy import displacy
from collections import Counter
nlp = spacy.load("en_core_web_lg")
df = pd.read_csv('restaurant_reviews_az.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48147 entries, 0 to 48146
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   review_id    48147 non-null  object
 1   user_id      48147 non-null  object
 2   business_id  48147 non-null  object
 3   stars        48147 non-null  int64 
 4   useful       48147 non-null  int64 
 5   funny        48147 non-null  int64 
 6   cool         48147 non-null  int64 
 7   text         48147 non-null  object
 8   date         48147 non-null  object
dtypes: int64(4), object(5)
memory usage: 3.3+ MB


* Imported all the libraries
* Loaded the trained model
* Read the dataset to df

## Code Cell 2 (5%) –Select the 1 star reviews and 5 star reviews from the dataset. 

In [2]:
df_selected = df[(df['stars']==1) | (df['stars']==5)]
df_selected.shape

(32933, 9)

In [3]:
df_selected.head(5)

review_id                 user_id             business_id  \
1  QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2  oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3  E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
5  kx6O_lyLzUnA7Xip5wh2NA  YsINprB2G1DM8qG1hbrPUg  rViAhfKLKmwbhTKROM9m0w   
6  DblKoOM1O6Bug_0b6YcpIQ  8o2iLbpduMiPefS2Gy_28g  wJmyu7W1K9A_gE8Ed4Bc9w   

   stars  useful  funny  cool  \
1      5       1      1     1   
2      5       1      0     0   
3      5       0      0     0   
5      1       0      0     0   
6      5       0      0     0   

                                                text                 date  
1  Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16  
2  I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44  
3  I've gone to claim Jumpers all over the US and...  2020-03-14 21:47:07  
5  I stay at the Main Hotel at the Casino from Ju...  2020-07-14 16:43:23  
6  In town after a long weekend of hiking and cam...  2020-01-20 04:55:52

In [4]:
# df_selected = df_selected.loc[:1500, ['text','stars']]
# df_selected

* There are total 32933 rows after filtering out the reviews with 1 and 5 stars

## Code Cell 3 (30%) - Apply necessary text processing techniques on the selected reviews. 

### Tokenization

In [5]:
df_selected['nlp_doc'] = df_selected['text'].apply(nlp)
df_selected

/var/folders/c4/kvf3th_j15lf5cq2pybvfgjm0000gn/T/ipykernel_1403/1815191854.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['nlp_doc'] = df_selected['text'].apply(nlp)


review_id                 user_id             business_id  \
1      QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2      oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3      E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
5      kx6O_lyLzUnA7Xip5wh2NA  YsINprB2G1DM8qG1hbrPUg  rViAhfKLKmwbhTKROM9m0w   
6      DblKoOM1O6Bug_0b6YcpIQ  8o2iLbpduMiPefS2Gy_28g  wJmyu7W1K9A_gE8Ed4Bc9w   
...                       ...                     ...                     ...   
48141  VkXr54yJMN4Qu4dRznAN8Q  4jEdEPDNAAa3aS7rYhQ60w  MK0OMY_u9unl8xSqjPLtMw   
48143  Zu4ng3tjf_2oa9LlnEvUmQ  ItQzeC91hJF6qvvE7-OZmQ  Yzh7Xo1_JBDWUl2BzRiYaQ   
48144  jacDcaIWSPdZq2bDq1GD_g  Y-mwrjOx29pnJX0MCBb2Yg  9VRmMY9vGhGKGz9hiGoEUw   
48145  dkGbETTcSQZTwHSnAMnLUw  _RmG_5kxRPgTWP7RptaFgQ  Bq0CQcwk5R8yhm-MGfHxCA   
48146  YGUyDcnAxZsUAjlK3QF99g  Qa5MBgb7Xyrr9Pg5bD4taQ  jh8j-DWqgWkbRe_a2XtKFQ   

       stars  useful  funny  cool  \
1          5       1      1     1   
2          5       1      0     0   
3          5       0      0     0   
5          1       0      0     0   
6          5       0      0     0   
...      ...     ...    ...   ...   
48141      5       3      0     0   
48143      5       0      0     0   
48144      1       0      0     0   
48145      5       6      2     4   
48146      5       0      0     1   

                                                    text                 date  \
1      Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16   
2      I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44   
3      I've gone to claim Jumpers all over the US and...  2020-03-14 21:47:07   
5      I stay at the Main Hotel at the Casino from Ju...  2020-07-14 16:43:23   
6      In town after a long weekend of hiking and cam...  2020-01-20 04:55:52   
...                                                  ...                  ...   
48141  ALWAYS love this place! And it's always busy, ...  2020-01-17 20:58:05   
48143  Fresh and delicious food, served fast. What mo...  2021-10-30 20:17:34   
48144  If I could leave no stars I would. I understan...  2021-11-28 14:23:39   
48145  Bosnian food?? \n\n--- location. This is a HID...  2020-01-05 04:20:42   
48146  My secret to finding good food often starts by...  2021-11-28 22:24:05   

                                                 nlp_doc  
1      (Pandemic, pit, stop, to, have, an, ice, cream...  
2      (I, was, lucky, enough, to, go, to, the, soft,...  
3      (I, 've, gone, to, claim, Jumpers, all, over, ...  
5      (I, stay, at, the, Main, Hotel, at, the, Casin...  
6      (In, town, after, a, long, weekend, of, hiking...  
...                                                  ...  
48141  (ALWAYS, love, this, place, !, And, it, 's, al...  
48143  (Fresh, and, delicious, food, ,, served, fast,...  
48144  (If, I, could, leave, no, stars, I, would, ., ...  
48145  (Bosnian, food, ?, ?, \n\n, ---, location, ., ...  
48146  (My, secret, to, finding, good, food, often, s...  

[32933 rows x 10 columns]

* Applied the pre trained model to text column

#### Sentence and Word Tokenization

In [6]:
df_selected['sentence'] = df_selected['nlp_doc'].apply(lambda x: [sent for sent in nlp(x).sents])
df_selected['word'] = df_selected['nlp_doc'].apply(lambda x: [token for token in nlp(x)])
df_selected

/var/folders/c4/kvf3th_j15lf5cq2pybvfgjm0000gn/T/ipykernel_1403/2079521788.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['sentence'] = df_selected['nlp_doc'].apply(lambda x: [sent for sent in nlp(x).sents])
/var/folders/c4/kvf3th_j15lf5cq2pybvfgjm0000gn/T/ipykernel_1403/2079521788.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['word'] = df_selected['nlp_doc'].apply(lambda x: [token for token in nlp(x)])


review_id                 user_id             business_id  \
1      QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2      oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3      E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
5      kx6O_lyLzUnA7Xip5wh2NA  YsINprB2G1DM8qG1hbrPUg  rViAhfKLKmwbhTKROM9m0w   
6      DblKoOM1O6Bug_0b6YcpIQ  8o2iLbpduMiPefS2Gy_28g  wJmyu7W1K9A_gE8Ed4Bc9w   
...                       ...                     ...                     ...   
48141  VkXr54yJMN4Qu4dRznAN8Q  4jEdEPDNAAa3aS7rYhQ60w  MK0OMY_u9unl8xSqjPLtMw   
48143  Zu4ng3tjf_2oa9LlnEvUmQ  ItQzeC91hJF6qvvE7-OZmQ  Yzh7Xo1_JBDWUl2BzRiYaQ   
48144  jacDcaIWSPdZq2bDq1GD_g  Y-mwrjOx29pnJX0MCBb2Yg  9VRmMY9vGhGKGz9hiGoEUw   
48145  dkGbETTcSQZTwHSnAMnLUw  _RmG_5kxRPgTWP7RptaFgQ  Bq0CQcwk5R8yhm-MGfHxCA   
48146  YGUyDcnAxZsUAjlK3QF99g  Qa5MBgb7Xyrr9Pg5bD4taQ  jh8j-DWqgWkbRe_a2XtKFQ   

       stars  useful  funny  cool  \
1          5       1      1     1   
2          5       1      0     0   
3          5       0      0     0   
5          1       0      0     0   
6          5       0      0     0   
...      ...     ...    ...   ...   
48141      5       3      0     0   
48143      5       0      0     0   
48144      1       0      0     0   
48145      5       6      2     4   
48146      5       0      0     1   

                                                    text                 date  \
1      Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16   
2      I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44   
3      I've gone to claim Jumpers all over the US and...  2020-03-14 21:47:07   
5      I stay at the Main Hotel at the Casino from Ju...  2020-07-14 16:43:23   
6      In town after a long weekend of hiking and cam...  2020-01-20 04:55:52   
...                                                  ...                  ...   
48141  ALWAYS love this place! And it's always busy, ...  2020-01-17 20:58:05   
48143  Fresh and delicious food, served fast. What mo...  2021-10-30 20:17:34   
48144  If I could leave no stars I would. I understan...  2021-11-28 14:23:39   
48145  Bosnian food?? \n\n--- location. This is a HID...  2020-01-05 04:20:42   
48146  My secret to finding good food often starts by...  2021-11-28 22:24:05   

                                                 nlp_doc  \
1      (Pandemic, pit, stop, to, have, an, ice, cream...   
2      (I, was, lucky, enough, to, go, to, the, soft,...   
3      (I, 've, gone, to, claim, Jumpers, all, over, ...   
5      (I, stay, at, the, Main, Hotel, at, the, Casin...   
6      (In, town, after, a, long, weekend, of, hiking...   
...                                                  ...   
48141  (ALWAYS, love, this, place, !, And, it, 's, al...   
48143  (Fresh, and, delicious, food, ,, served, fast,...   
48144  (If, I, could, leave, no, stars, I, would, ., ...   
48145  (Bosnian, food, ?, ?, \n\n, ---, location, ., ...   
48146  (My, secret, to, finding, good, food, often, s...   

                                                sentence  \
1      [(Pandemic, pit, stop, to, have, an, ice, crea...   
2      [(I, was, lucky, enough, to, go, to, the, soft...   
3      [(I, 've, gone, to, claim, Jumpers, all, over,...   
5      [(I, stay, at, the, Main, Hotel, at, the, Casi...   
6      [(In, town, after, a, long, weekend, of, hikin...   
...                                                  ...   
48141  [(ALWAYS, love, this, place, !), (And, it, 's,...   
48143  [(Fresh, and, delicious, food, ,, served, fast...   
48144  [(If, I, could, leave, no, stars, I, would, .)...   
48145  [(Bosnian, food, ?, ?, \n\n, ---, location, .)...   
48146  [(My, secret, to, finding, good, food, often, ...   

                                                    word  
1      [Pandemic, pit, stop, to, have, an, ice, cream...  
2      [I, was, lucky, enough, to, go, to, the, soft,...  
3      [I,

* Performed Sentence Tokentisation
* Performed Word Tokenization

#### Stop Words Removal

In [7]:
df_selected['cleaned_words'] = df_selected['word'].apply(lambda tokens: [token for token in tokens if not token.is_stop])
df_selected

/var/folders/c4/kvf3th_j15lf5cq2pybvfgjm0000gn/T/ipykernel_1403/1973501910.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['cleaned_words'] = df_selected['word'].apply(lambda tokens: [token for token in tokens if not token.is_stop])


review_id                 user_id             business_id  \
1      QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2      oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3      E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
5      kx6O_lyLzUnA7Xip5wh2NA  YsINprB2G1DM8qG1hbrPUg  rViAhfKLKmwbhTKROM9m0w   
6      DblKoOM1O6Bug_0b6YcpIQ  8o2iLbpduMiPefS2Gy_28g  wJmyu7W1K9A_gE8Ed4Bc9w   
...                       ...                     ...                     ...   
48141  VkXr54yJMN4Qu4dRznAN8Q  4jEdEPDNAAa3aS7rYhQ60w  MK0OMY_u9unl8xSqjPLtMw   
48143  Zu4ng3tjf_2oa9LlnEvUmQ  ItQzeC91hJF6qvvE7-OZmQ  Yzh7Xo1_JBDWUl2BzRiYaQ   
48144  jacDcaIWSPdZq2bDq1GD_g  Y-mwrjOx29pnJX0MCBb2Yg  9VRmMY9vGhGKGz9hiGoEUw   
48145  dkGbETTcSQZTwHSnAMnLUw  _RmG_5kxRPgTWP7RptaFgQ  Bq0CQcwk5R8yhm-MGfHxCA   
48146  YGUyDcnAxZsUAjlK3QF99g  Qa5MBgb7Xyrr9Pg5bD4taQ  jh8j-DWqgWkbRe_a2XtKFQ   

       stars  useful  funny  cool  \
1          5       1      1     1   
2          5       1      0     0   
3          5       0      0     0   
5          1       0      0     0   
6          5       0      0     0   
...      ...     ...    ...   ...   
48141      5       3      0     0   
48143      5       0      0     0   
48144      1       0      0     0   
48145      5       6      2     4   
48146      5       0      0     1   

                                                    text                 date  \
1      Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16   
2      I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44   
3      I've gone to claim Jumpers all over the US and...  2020-03-14 21:47:07   
5      I stay at the Main Hotel at the Casino from Ju...  2020-07-14 16:43:23   
6      In town after a long weekend of hiking and cam...  2020-01-20 04:55:52   
...                                                  ...                  ...   
48141  ALWAYS love this place! And it's always busy, ...  2020-01-17 20:58:05   
48143  Fresh and delicious food, served fast. What mo...  2021-10-30 20:17:34   
48144  If I could leave no stars I would. I understan...  2021-11-28 14:23:39   
48145  Bosnian food?? \n\n--- location. This is a HID...  2020-01-05 04:20:42   
48146  My secret to finding good food often starts by...  2021-11-28 22:24:05   

                                                 nlp_doc  \
1      (Pandemic, pit, stop, to, have, an, ice, cream...   
2      (I, was, lucky, enough, to, go, to, the, soft,...   
3      (I, 've, gone, to, claim, Jumpers, all, over, ...   
5      (I, stay, at, the, Main, Hotel, at, the, Casin...   
6      (In, town, after, a, long, weekend, of, hiking...   
...                                                  ...   
48141  (ALWAYS, love, this, place, !, And, it, 's, al...   
48143  (Fresh, and, delicious, food, ,, served, fast,...   
48144  (If, I, could, leave, no, stars, I, would, ., ...   
48145  (Bosnian, food, ?, ?, \n\n, ---, location, ., ...   
48146  (My, secret, to, finding, good, food, often, s...   

                                                sentence  \
1      [(Pandemic, pit, stop, to, have, an, ice, crea...   
2      [(I, was, lucky, enough, to, go, to, the, soft...   
3      [(I, 've, gone, to, claim, Jumpers, all, over,...   
5      [(I, stay, at, the, Main, Hotel, at, the, Casi...   
6      [(In, town, after, a, long, weekend, of, hikin...   
...                                                  ...   
48141  [(ALWAYS, love, this, place, !), (And, it, 's,...   
48143  [(Fresh, and, delicious, food, ,, served, fast...   
48144  [(If, I, could, leave, no, stars, I, would, .)...   
48145  [(Bosnian, food, ?, ?, \n\n, ---, location, .)...   
48146  [(My, secret, to, finding, good, food, often, ...   

                                                    word  \
1      [Pandemic, pit, stop, to, have, an, ice, cream...   
2      [I, was, lucky, enough, to, go, to, the, soft,...   
3      

* Removed the stop words

### Part-Of-Speech Tagging

In [8]:
df_selected['pos'] = df_selected['nlp_doc'].apply(lambda doc: [(token.text, token.pos_) for token in doc])
df_selected

/var/folders/c4/kvf3th_j15lf5cq2pybvfgjm0000gn/T/ipykernel_1403/2345859219.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['pos'] = df_selected['nlp_doc'].apply(lambda doc: [(token.text, token.pos_) for token in doc])


review_id                 user_id             business_id  \
1      QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2      oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3      E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
5      kx6O_lyLzUnA7Xip5wh2NA  YsINprB2G1DM8qG1hbrPUg  rViAhfKLKmwbhTKROM9m0w   
6      DblKoOM1O6Bug_0b6YcpIQ  8o2iLbpduMiPefS2Gy_28g  wJmyu7W1K9A_gE8Ed4Bc9w   
...                       ...                     ...                     ...   
48141  VkXr54yJMN4Qu4dRznAN8Q  4jEdEPDNAAa3aS7rYhQ60w  MK0OMY_u9unl8xSqjPLtMw   
48143  Zu4ng3tjf_2oa9LlnEvUmQ  ItQzeC91hJF6qvvE7-OZmQ  Yzh7Xo1_JBDWUl2BzRiYaQ   
48144  jacDcaIWSPdZq2bDq1GD_g  Y-mwrjOx29pnJX0MCBb2Yg  9VRmMY9vGhGKGz9hiGoEUw   
48145  dkGbETTcSQZTwHSnAMnLUw  _RmG_5kxRPgTWP7RptaFgQ  Bq0CQcwk5R8yhm-MGfHxCA   
48146  YGUyDcnAxZsUAjlK3QF99g  Qa5MBgb7Xyrr9Pg5bD4taQ  jh8j-DWqgWkbRe_a2XtKFQ   

       stars  useful  funny  cool  \
1          5       1      1     1   
2          5       1      0     0   
3          5       0      0     0   
5          1       0      0     0   
6          5       0      0     0   
...      ...     ...    ...   ...   
48141      5       3      0     0   
48143      5       0      0     0   
48144      1       0      0     0   
48145      5       6      2     4   
48146      5       0      0     1   

                                                    text                 date  \
1      Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16   
2      I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44   
3      I've gone to claim Jumpers all over the US and...  2020-03-14 21:47:07   
5      I stay at the Main Hotel at the Casino from Ju...  2020-07-14 16:43:23   
6      In town after a long weekend of hiking and cam...  2020-01-20 04:55:52   
...                                                  ...                  ...   
48141  ALWAYS love this place! And it's always busy, ...  2020-01-17 20:58:05   
48143  Fresh and delicious food, served fast. What mo...  2021-10-30 20:17:34   
48144  If I could leave no stars I would. I understan...  2021-11-28 14:23:39   
48145  Bosnian food?? \n\n--- location. This is a HID...  2020-01-05 04:20:42   
48146  My secret to finding good food often starts by...  2021-11-28 22:24:05   

                                                 nlp_doc  \
1      (Pandemic, pit, stop, to, have, an, ice, cream...   
2      (I, was, lucky, enough, to, go, to, the, soft,...   
3      (I, 've, gone, to, claim, Jumpers, all, over, ...   
5      (I, stay, at, the, Main, Hotel, at, the, Casin...   
6      (In, town, after, a, long, weekend, of, hiking...   
...                                                  ...   
48141  (ALWAYS, love, this, place, !, And, it, 's, al...   
48143  (Fresh, and, delicious, food, ,, served, fast,...   
48144  (If, I, could, leave, no, stars, I, would, ., ...   
48145  (Bosnian, food, ?, ?, \n\n, ---, location, ., ...   
48146  (My, secret, to, finding, good, food, often, s...   

                                                sentence  \
1      [(Pandemic, pit, stop, to, have, an, ice, crea...   
2      [(I, was, lucky, enough, to, go, to, the, soft...   
3      [(I, 've, gone, to, claim, Jumpers, all, over,...   
5      [(I, stay, at, the, Main, Hotel, at, the, Casi...   
6      [(In, town, after, a, long, weekend, of, hikin...   
...                                                  ...   
48141  [(ALWAYS, love, this, place, !), (And, it, 's,...   
48143  [(Fresh, and, delicious, food, ,, served, fast...   
48144  [(If, I, could, leave, no, stars, I, would, .)...   
48145  [(Bosnian, food, ?, ?, \n\n, ---, location, .)...   
48146  [(My, secret, to, finding, good, food, often, ...   

                                                    word  \
1      [Pandemic, pit, stop, to, have, an, ice, cream...   
2      [I, was, lucky, enough, to, go, to, the, soft,...   
3      

* Performed POS tagging and saved it pos column in the dataframe

### Lemmatization

In [9]:
df_selected['lemmatized'] = df_selected['nlp_doc'].apply(lambda doc: [token.lemma_ for token in doc])
df_selected

/var/folders/c4/kvf3th_j15lf5cq2pybvfgjm0000gn/T/ipykernel_1403/721325772.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['lemmatized'] = df_selected['nlp_doc'].apply(lambda doc: [token.lemma_ for token in doc])


review_id                 user_id             business_id  \
1      QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2      oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3      E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
5      kx6O_lyLzUnA7Xip5wh2NA  YsINprB2G1DM8qG1hbrPUg  rViAhfKLKmwbhTKROM9m0w   
6      DblKoOM1O6Bug_0b6YcpIQ  8o2iLbpduMiPefS2Gy_28g  wJmyu7W1K9A_gE8Ed4Bc9w   
...                       ...                     ...                     ...   
48141  VkXr54yJMN4Qu4dRznAN8Q  4jEdEPDNAAa3aS7rYhQ60w  MK0OMY_u9unl8xSqjPLtMw   
48143  Zu4ng3tjf_2oa9LlnEvUmQ  ItQzeC91hJF6qvvE7-OZmQ  Yzh7Xo1_JBDWUl2BzRiYaQ   
48144  jacDcaIWSPdZq2bDq1GD_g  Y-mwrjOx29pnJX0MCBb2Yg  9VRmMY9vGhGKGz9hiGoEUw   
48145  dkGbETTcSQZTwHSnAMnLUw  _RmG_5kxRPgTWP7RptaFgQ  Bq0CQcwk5R8yhm-MGfHxCA   
48146  YGUyDcnAxZsUAjlK3QF99g  Qa5MBgb7Xyrr9Pg5bD4taQ  jh8j-DWqgWkbRe_a2XtKFQ   

       stars  useful  funny  cool  \
1          5       1      1     1   
2          5       1      0     0   
3          5       0      0     0   
5          1       0      0     0   
6          5       0      0     0   
...      ...     ...    ...   ...   
48141      5       3      0     0   
48143      5       0      0     0   
48144      1       0      0     0   
48145      5       6      2     4   
48146      5       0      0     1   

                                                    text                 date  \
1      Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16   
2      I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44   
3      I've gone to claim Jumpers all over the US and...  2020-03-14 21:47:07   
5      I stay at the Main Hotel at the Casino from Ju...  2020-07-14 16:43:23   
6      In town after a long weekend of hiking and cam...  2020-01-20 04:55:52   
...                                                  ...                  ...   
48141  ALWAYS love this place! And it's always busy, ...  2020-01-17 20:58:05   
48143  Fresh and delicious food, served fast. What mo...  2021-10-30 20:17:34   
48144  If I could leave no stars I would. I understan...  2021-11-28 14:23:39   
48145  Bosnian food?? \n\n--- location. This is a HID...  2020-01-05 04:20:42   
48146  My secret to finding good food often starts by...  2021-11-28 22:24:05   

                                                 nlp_doc  \
1      (Pandemic, pit, stop, to, have, an, ice, cream...   
2      (I, was, lucky, enough, to, go, to, the, soft,...   
3      (I, 've, gone, to, claim, Jumpers, all, over, ...   
5      (I, stay, at, the, Main, Hotel, at, the, Casin...   
6      (In, town, after, a, long, weekend, of, hiking...   
...                                                  ...   
48141  (ALWAYS, love, this, place, !, And, it, 's, al...   
48143  (Fresh, and, delicious, food, ,, served, fast,...   
48144  (If, I, could, leave, no, stars, I, would, ., ...   
48145  (Bosnian, food, ?, ?, \n\n, ---, location, ., ...   
48146  (My, secret, to, finding, good, food, often, s...   

                                                sentence  \
1      [(Pandemic, pit, stop, to, have, an, ice, crea...   
2      [(I, was, lucky, enough, to, go, to, the, soft...   
3      [(I, 've, gone, to, claim, Jumpers, all, over,...   
5      [(I, stay, at, the, Main, Hotel, at, the, Casi...   
6      [(In, town, after, a, long, weekend, of, hikin...   
...                                                  ...   
48141  [(ALWAYS, love, this, place, !), (And, it, 's,...   
48143  [(Fresh, and, delicious, food, ,, served, fast...   
48144  [(If, I, could, leave, no, stars, I, would, .)...   
48145  [(Bosnian, food, ?, ?, \n\n, ---, location, .)...   
48146  [(My, secret, to, finding, good, food, often, ...   

                                                    word  \
1      [Pandemic, pit, stop, to, have, an, ice, cream...   
2      [I, was, lucky, enough, to, go, to, the, soft,...   
3      

* Performed Lemmatization and saved it to a new column

### Names Entity Recognization

In [10]:
df_selected['entities'] = df_selected['nlp_doc'].apply(lambda doc: [(ent.text, ent.label_) for ent in doc.ents])
df_selected

/var/folders/c4/kvf3th_j15lf5cq2pybvfgjm0000gn/T/ipykernel_1403/327110970.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['entities'] = df_selected['nlp_doc'].apply(lambda doc: [(ent.text, ent.label_) for ent in doc.ents])


review_id                 user_id             business_id  \
1      QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2      oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3      E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
5      kx6O_lyLzUnA7Xip5wh2NA  YsINprB2G1DM8qG1hbrPUg  rViAhfKLKmwbhTKROM9m0w   
6      DblKoOM1O6Bug_0b6YcpIQ  8o2iLbpduMiPefS2Gy_28g  wJmyu7W1K9A_gE8Ed4Bc9w   
...                       ...                     ...                     ...   
48141  VkXr54yJMN4Qu4dRznAN8Q  4jEdEPDNAAa3aS7rYhQ60w  MK0OMY_u9unl8xSqjPLtMw   
48143  Zu4ng3tjf_2oa9LlnEvUmQ  ItQzeC91hJF6qvvE7-OZmQ  Yzh7Xo1_JBDWUl2BzRiYaQ   
48144  jacDcaIWSPdZq2bDq1GD_g  Y-mwrjOx29pnJX0MCBb2Yg  9VRmMY9vGhGKGz9hiGoEUw   
48145  dkGbETTcSQZTwHSnAMnLUw  _RmG_5kxRPgTWP7RptaFgQ  Bq0CQcwk5R8yhm-MGfHxCA   
48146  YGUyDcnAxZsUAjlK3QF99g  Qa5MBgb7Xyrr9Pg5bD4taQ  jh8j-DWqgWkbRe_a2XtKFQ   

       stars  useful  funny  cool  \
1          5       1      1     1   
2          5       1      0     0   
3          5       0      0     0   
5          1       0      0     0   
6          5       0      0     0   
...      ...     ...    ...   ...   
48141      5       3      0     0   
48143      5       0      0     0   
48144      1       0      0     0   
48145      5       6      2     4   
48146      5       0      0     1   

                                                    text                 date  \
1      Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16   
2      I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44   
3      I've gone to claim Jumpers all over the US and...  2020-03-14 21:47:07   
5      I stay at the Main Hotel at the Casino from Ju...  2020-07-14 16:43:23   
6      In town after a long weekend of hiking and cam...  2020-01-20 04:55:52   
...                                                  ...                  ...   
48141  ALWAYS love this place! And it's always busy, ...  2020-01-17 20:58:05   
48143  Fresh and delicious food, served fast. What mo...  2021-10-30 20:17:34   
48144  If I could leave no stars I would. I understan...  2021-11-28 14:23:39   
48145  Bosnian food?? \n\n--- location. This is a HID...  2020-01-05 04:20:42   
48146  My secret to finding good food often starts by...  2021-11-28 22:24:05   

                                                 nlp_doc  \
1      (Pandemic, pit, stop, to, have, an, ice, cream...   
2      (I, was, lucky, enough, to, go, to, the, soft,...   
3      (I, 've, gone, to, claim, Jumpers, all, over, ...   
5      (I, stay, at, the, Main, Hotel, at, the, Casin...   
6      (In, town, after, a, long, weekend, of, hiking...   
...                                                  ...   
48141  (ALWAYS, love, this, place, !, And, it, 's, al...   
48143  (Fresh, and, delicious, food, ,, served, fast,...   
48144  (If, I, could, leave, no, stars, I, would, ., ...   
48145  (Bosnian, food, ?, ?, \n\n, ---, location, ., ...   
48146  (My, secret, to, finding, good, food, often, s...   

                                                sentence  \
1      [(Pandemic, pit, stop, to, have, an, ice, crea...   
2      [(I, was, lucky, enough, to, go, to, the, soft...   
3      [(I, 've, gone, to, claim, Jumpers, all, over,...   
5      [(I, stay, at, the, Main, Hotel, at, the, Casi...   
6      [(In, town, after, a, long, weekend, of, hikin...   
...                                                  ...   
48141  [(ALWAYS, love, this, place, !), (And, it, 's,...   
48143  [(Fresh, and, delicious, food, ,, served, fast...   
48144  [(If, I, could, leave, no, stars, I, would, .)...   
48145  [(Bosnian, food, ?, ?, \n\n, ---, location, .)...   
48146  [(My, secret, to, finding, good, food, often, ...   

                                                    word  \
1      [Pandemic, pit, stop, to, have, an, ice, cream...   
2      [I, was, lucky, enough, to, go, to, the, soft,...   
3      

* Performed NER and saved it to a new column entities in the dataframe

In [11]:
displacy.render(df_selected['nlp_doc'].iloc[0], style='ent', jupyter=True)

### Parsing

In [12]:
df_selected['parsing'] = df_selected['nlp_doc'].apply(lambda doc: [(token.text, token.dep_, token.head.text) for token in doc])
df_selected

/var/folders/c4/kvf3th_j15lf5cq2pybvfgjm0000gn/T/ipykernel_1403/4069891031.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['parsing'] = df_selected['nlp_doc'].apply(lambda doc: [(token.text, token.dep_, token.head.text) for token in doc])


review_id                 user_id             business_id  \
1      QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2      oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3      E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
5      kx6O_lyLzUnA7Xip5wh2NA  YsINprB2G1DM8qG1hbrPUg  rViAhfKLKmwbhTKROM9m0w   
6      DblKoOM1O6Bug_0b6YcpIQ  8o2iLbpduMiPefS2Gy_28g  wJmyu7W1K9A_gE8Ed4Bc9w   
...                       ...                     ...                     ...   
48141  VkXr54yJMN4Qu4dRznAN8Q  4jEdEPDNAAa3aS7rYhQ60w  MK0OMY_u9unl8xSqjPLtMw   
48143  Zu4ng3tjf_2oa9LlnEvUmQ  ItQzeC91hJF6qvvE7-OZmQ  Yzh7Xo1_JBDWUl2BzRiYaQ   
48144  jacDcaIWSPdZq2bDq1GD_g  Y-mwrjOx29pnJX0MCBb2Yg  9VRmMY9vGhGKGz9hiGoEUw   
48145  dkGbETTcSQZTwHSnAMnLUw  _RmG_5kxRPgTWP7RptaFgQ  Bq0CQcwk5R8yhm-MGfHxCA   
48146  YGUyDcnAxZsUAjlK3QF99g  Qa5MBgb7Xyrr9Pg5bD4taQ  jh8j-DWqgWkbRe_a2XtKFQ   

       stars  useful  funny  cool  \
1          5       1      1     1   
2          5       1      0     0   
3          5       0      0     0   
5          1       0      0     0   
6          5       0      0     0   
...      ...     ...    ...   ...   
48141      5       3      0     0   
48143      5       0      0     0   
48144      1       0      0     0   
48145      5       6      2     4   
48146      5       0      0     1   

                                                    text                 date  \
1      Pandemic pit stop to have an ice cream.... onl...  2020-04-19 05:33:16   
2      I was lucky enough to go to the soft opening a...  2020-02-29 19:43:44   
3      I've gone to claim Jumpers all over the US and...  2020-03-14 21:47:07   
5      I stay at the Main Hotel at the Casino from Ju...  2020-07-14 16:43:23   
6      In town after a long weekend of hiking and cam...  2020-01-20 04:55:52   
...                                                  ...                  ...   
48141  ALWAYS love this place! And it's always busy, ...  2020-01-17 20:58:05   
48143  Fresh and delicious food, served fast. What mo...  2021-10-30 20:17:34   
48144  If I could leave no stars I would. I understan...  2021-11-28 14:23:39   
48145  Bosnian food?? \n\n--- location. This is a HID...  2020-01-05 04:20:42   
48146  My secret to finding good food often starts by...  2021-11-28 22:24:05   

                                                 nlp_doc  \
1      (Pandemic, pit, stop, to, have, an, ice, cream...   
2      (I, was, lucky, enough, to, go, to, the, soft,...   
3      (I, 've, gone, to, claim, Jumpers, all, over, ...   
5      (I, stay, at, the, Main, Hotel, at, the, Casin...   
6      (In, town, after, a, long, weekend, of, hiking...   
...                                                  ...   
48141  (ALWAYS, love, this, place, !, And, it, 's, al...   
48143  (Fresh, and, delicious, food, ,, served, fast,...   
48144  (If, I, could, leave, no, stars, I, would, ., ...   
48145  (Bosnian, food, ?, ?, \n\n, ---, location, ., ...   
48146  (My, secret, to, finding, good, food, often, s...   

                                                sentence  \
1      [(Pandemic, pit, stop, to, have, an, ice, crea...   
2      [(I, was, lucky, enough, to, go, to, the, soft...   
3      [(I, 've, gone, to, claim, Jumpers, all, over,...   
5      [(I, stay, at, the, Main, Hotel, at, the, Casi...   
6      [(In, town, after, a, long, weekend, of, hikin...   
...                                                  ...   
48141  [(ALWAYS, love, this, place, !), (And, it, 's,...   
48143  [(Fresh, and, delicious, food, ,, served, fast...   
48144  [(If, I, could, leave, no, stars, I, would, .)...   
48145  [(Bosnian, food, ?, ?, \n\n, ---, location, .)...   
48146  [(My, secret, to, finding, good, food, often, ...   

                                                    word  \
1      [Pandemic, pit, stop, to, have, an, ice, cream...   
2      [I, was, lucky, enough, to, go, to, the, soft,...   
3      

* Performed Parsing and saved it to a new column parsing

## Code Cell 4 (10%) – Find the top 20 frequently used nouns in 1 star reviews and 5 star reviews, respectively.

In [13]:
noun = {}

for i in df_selected[df_selected['stars'] == 1]['pos']:
    length = len(i)
    for j in range(length):
        if i[j][1] == "NOUN":
            if i[j][0] in noun:
                noun[i[j][0]] += 1
            else:
                noun[i[j][0]] = 1

top_20 = sorted(noun.items(), key=lambda x: x[1], reverse=True)[:20]
print("Top 20 nouns for 1 Star Restaurants:")
for i,j in top_20:
    print(f"{i}: {j}")

Top 20 nouns for 1 Star Restaurants:
food: 6133
order: 3673
time: 3375
place: 2935
service: 2840
minutes: 2381
restaurant: 1999
manager: 1477
people: 1384
customer: 1263
chicken: 1252
table: 1207
location: 1103
experience: 1038
pizza: 1032
staff: 980
way: 933
times: 932
server: 881
meal: 873


In [14]:
noun = {}

for i in df_selected[df_selected['stars'] == 5]['pos']:
    length = len(i)
    for j in range(length):
        if i[j][1] == "NOUN":
            if i[j][0] in noun:
                noun[i[j][0]] += 1
            else:
                noun[i[j][0]] = 1

top_20 = sorted(noun.items(), key=lambda x: x[1], reverse=True)[:20]
print("Top 20 nouns for 5 Star Restaurants:")
for i,j in top_20:
    print(f"{i}: {j}")

Top 20 nouns for 5 Star Restaurants:
food: 14658
place: 9698
service: 6569
time: 5554
restaurant: 3934
staff: 3688
menu: 3091
chicken: 2825
pizza: 2465
experience: 2103
sauce: 1970
meal: 1906
order: 1848
breakfast: 1819
tacos: 1804
lunch: 1797
salad: 1788
spot: 1713
cheese: 1655
family: 1652


* Saved all the count of the noun words of how many times they have been used in the reviews in dictionary.
* Printed out Top 20 nouns for both 1 star and 5 star reviews.

## Code Cell 5 (10%) – Find the top 20 frequently used adjectives in  1 star reviews and 5 star, respectively.

In [15]:
adj = {}

for i in df_selected[df_selected['stars'] == 1]['pos']:
    length = len(i)
    for j in range(length):
        if i[j][1] == "ADJ":
            if i[j][0] in adj:
                adj[i[j][0]] += 1
            else:
                adj[i[j][0]] = 1

top_20 = sorted(adj.items(), key=lambda x: x[1], reverse=True)[:20]
print("Top 20 adjectives for 1 Star Restaurants:")
for i,j in top_20:
    print(f"{i}: {j}")

Top 20 adjectives for 1 Star Restaurants:
good: 1782
other: 1340
bad: 1138
rude: 988
cold: 866
more: 825
last: 814
first: 747
worst: 692
great: 672
wrong: 656
better: 654
same: 607
sure: 598
disappointed: 567
terrible: 559
new: 550
many: 528
horrible: 520
few: 501


In [16]:
adj = {}

for i in df_selected[df_selected['stars'] == 5]['pos']:
    length = len(i)
    for j in range(length):
        if i[j][1] == "ADJ":
            if i[j][0] in adj:
                adj[i[j][0]] += 1
            else:
                adj[i[j][0]] = 1

top_20 = sorted(adj.items(), key=lambda x: x[1], reverse=True)[:20]
print("Top 20 adjectives for 5 Star Restaurants:")
for i,j in top_20:
    print(f"{i}: {j}")

Top 20 adjectives for 5 Star Restaurants:
great: 9366
good: 8928
delicious: 6547
amazing: 4673
best: 4268
friendly: 4041
fresh: 3170
nice: 2962
Great: 2809
favorite: 2144
more: 1983
excellent: 1956
little: 1951
other: 1932
perfect: 1931
first: 1730
hot: 1675
tasty: 1523
new: 1517
sure: 1495


* Saved all the count of the ajective words of how many times they have been used in the reviews in dictionary.
* Printed out Top 20 ajective for both 1 star and 5 star reviews.

## Code Cell 6 (10%) – Find the top 20 frequently used verbs in 1 star reviews and 5 star reviews, respectively.


In [17]:
verb = {}

for i in df_selected[df_selected['stars'] == 1]['pos']:
    length = len(i)
    for j in range(length):
        if i[j][1] == "VERB":
            if i[j][0] in verb:
                verb[i[j][0]] += 1
            else:
                verb[i[j][0]] = 1

top_20 = sorted(verb.items(), key=lambda x: x[1], reverse=True)[:20]
print("Top 20 verbs for 1 Star Restaurants:")
for i,j in top_20:
    print(f"{i}: {j}")

Top 20 verbs for 1 Star Restaurants:
had: 3919
get: 2929
have: 2565
ordered: 2473
go: 2416
said: 2361
got: 2065
asked: 1933
told: 1871
came: 1571
went: 1453
going: 1301
know: 1252
give: 1159
come: 1136
eat: 1110
take: 1079
make: 1032
took: 1030
order: 1005


In [18]:
verb = {}

for i in df_selected[df_selected['stars'] == 5]['pos']:
    length = len(i)
    for j in range(length):
        if i[j][1] == "VERB":
            if i[j][0] in verb:
                verb[i[j][0]] += 1
            else:
                verb[i[j][0]] = 1

top_20 = sorted(verb.items(), key=lambda x: x[1], reverse=True)[:20]
print("Top 20 verbs for 5 Star Restaurants:")
for i,j in top_20:
    print(f"{i}: {j}")

Top 20 verbs for 5 Star Restaurants:
had: 10521
have: 5759
go: 4071
get: 3690
ordered: 3548
got: 3378
try: 3139
love: 3025
recommend: 2784
made: 2621
come: 2235
eat: 2005
came: 1941
make: 1788
take: 1706
tried: 1693
order: 1658
has: 1621
went: 1497
going: 1421


* Saved all the count of the verb words of how many times they have been used in the reviews in dictionary.
* Printed out Top 20 verb for both 1 star and 5 star reviews.

## Code Cell 7 (10%) – Find the top 20 frequently used named entities from the selected reviews.


In [19]:
all_entities = [ent for sublist in df_selected['entities'] for ent in sublist]

entity_counter = Counter([ent[0] for ent in all_entities])

top_20_entities = entity_counter.most_common(20)

print("Top 20 named entities:")
for entity, count in top_20_entities:
    print(f"{entity}: {count}")

Top 20 named entities:
Tucson: 5385
first: 2850
one: 2169
two: 2081
2: 1501
Mexican: 1373
today: 1097
3: 1081
5: 1037
One: 848
First: 848
4: 706
1: 638
half: 636
three: 624
second: 613
Italian: 588
tonight: 551
Chinese: 517
Thai: 454


* Printed out Top 20 named entities using the counter function from collection library

## Text Cell 8 (10%) - Write one paragraph about your observation on the language of 1 star and 5 star reviews. What is the key to a good restaurant experience? 

 I noticed that 1-star reviews often use negative and critical terms like "terrible" and "rude," reflecting dissatisfaction with service, food quality, or value. On the other hand, 5-star reviews feature positive and enthusiastic language, with words like "amazing" and "friendly," showcasing satisfaction with excellent food, attentive service, and cleanliness. From this, I believe the key to a great restaurant experience lies in consistently delivering high-quality meals, exceptional service, and a welcoming atmosphere.

## Text Cell 9 (5%) - Acknowledge if you have used any GenAI tools in this assignment and anyone you have worked together with on this assignment. 

I referred to the code file uploaded by professor which was taught in class which helped me a lot in building the intuition. Additonally, I took help from chatgpt for fixing syntax in a few code lines.